sqlalchemy-stubs/sql/ddl.pyi

```diff
@@ -1,117 +1,143 @@
-from typing import Any, Optional
-from .elements import ClauseElement as ClauseElement
+from typing import Any, Optional, Union, List, Dict, TypeVar, Generic, Sequence, Callable, Tuple
+from typing_extensions import Protocol
+from .elements import ClauseElement as ClauseElement, ColumnElement
 from .base import Executable as Executable, SchemaVisitor as SchemaVisitor
+from .schema import Table, MetaData, SchemaItem, ForeignKeyConstraint, ForeignKey
+from .compiler import IdentifierPreparer
+from .type_api import TypeEngine
+from .. import engine
+
+_T = TypeVar('_T')
 
 class _DDLCompiles(ClauseElement): ...
 
+class _DDLCallable(Protocol):
+    def __call__(self, ddl: DDLElement, target: Optional[Union[Table, MetaData]], bind: engine.Connection,
+                 tables: Optional[List[Any]] = ..., state: Optional[Any] = ...,
+                 checkfirst: bool = ...) -> bool: ...
+
+_DDLE = TypeVar('_DDLE', bound=DDLElement)
+
 class DDLElement(Executable, _DDLCompiles):
-    target: Any = ...
-    on: Any = ...
-    dialect: Any = ...
-    callable_: Any = ...
-    def execute(self, *multiparams, **params): ...
-    def execute_at(self, event_name, target): ...
-    def against(self, target): ...
+    target: Optional[SchemaItem] = ...
+    on: Optional[Union[str, Tuple[str, ...], _DDLOnCallback]] = ...
+    dialect: Optional[engine.Dialect] = ...
+    callable_: Optional[_DDLCallable] = ...
+    bind: Optional[engine.Connection] = ...
+    def execute(self, bind: Optional[engine.Connectable] = ...,  # type: ignore
+                target: Optional[SchemaItem] = ...) -> Optional[engine.ResultProxy]: ...
+    def execute_at(self, event_name: str, target: SchemaItem) -> None: ...
+    def against(self: _DDLE, target: SchemaItem) -> _DDLE: ...
     state: Any = ...
-    def execute_if(self, dialect: Optional[Any] = ..., callable_: Optional[Any] = ...,
-                   state: Optional[Any] = ...): ...
+    def execute_if(self: _DDLE, dialect: Optional[engine.Dialect] = ..., callable_: Optional[_DDLCallable] = ...,
+                   state: Optional[Any] = ...) -> _DDLE: ...
     def __call__(self, target, bind, **kw): ...
-    def bind(self): ...
+
+class _DDLOnCallback(Protocol):
+    def __call__(self, ddl: DDLElement, event: Optional[str], target: Optional[Union[Table, MetaData]],
+                 connection: engine.Connection, tables: Optional[List[Any]] = ...) -> bool: ...
```

Comment:
Maybe define both protocols together at the top of module? Or is this intentional?